### ISSUE TITLE
Redundant explicit nil check on variable makes it report nilable again

### ISSUE THREAD

**Issue by Luka967** [EmmyLuaLs/emmylua-analyzer-rust#1141]
May be a symptom of #1140, but reporting separately just in case it isn't.

```lua
local pure_condition --- @type boolean
--- @class Foo
--- @field next? Foo

local target --- @type Foo
local source --- @type Foo
if not source.next then return end
if source.next and pure_condition then return end
target = source.next
```
At assignment `source.next` is reported as `Foo` as expected

```lua
local target --- @type Foo
local source --- @type Foo
if source.next == nil then return end -- explicit nil check
if pure_condition then return end -- no redundant explicit nil check
target = source.next
```
At assignment `source.next` is reported as `Foo` as expected

```lua
local target --- @type Foo
local source --- @type Foo
if source.next == nil then return end -- explicit nil check
if source.next ~= nil and pure_condition then return end -- redundant explicit nil check
target = source.next
```
At assignment `source.next` is reported as `Foo?` and equivalent warning is emitted on it (Cannot assign `Foo?` to `Foo`)

emmylua_ls `v0.23.2`
VSCode `1.1.126`
LS integration on VSCode, `tangzx.emmylua`, version `0.9.39`
